src/history/html5.js

```diff
@@ -8,24 +8,37 @@ import { setupScroll, handleScroll } from '../util/scroll'
 import { pushState, replaceState, supportsPushState } from '../util/push-state'
 
 export class HTML5History extends History {
+  +initLocation: string
+
   constructor (router: Router, base: ?string) {
     super(router, base)
 
+    this.initLocation = getLocation(this.base)
+  }
+
+  setupListeners () {
+    if (this.listeners.length > 0) {
+      return
+    }
+
+    const router = this.router
     const expectScroll = router.options.scrollBehavior
     const supportsScroll = supportsPushState && expectScroll
 
     if (supportsScroll) {
-      setupScroll()
+      this.listeners.push(setupScroll())
     }
 
-    const initLocation = getLocation(this.base)
```

Comment:
I think we should still consider the initial location when listeners are attached and avoid storing a property `initLocation` on the instance. Were you thinking about something in particular when you added it?

Reply:
Hmm good question - I was under the impression that the "async guard" and "first popstate event" described on lines 38 and 39 were something that occurred only during initial page load, and not during subsequent navigation events.

But rethinking this, I guess the if statement on line 41 exists as a way to short circuit when we know that `transitionTo()` is going to fire during the `init` function in index.js?

Am happy to take direction on this one, as I'm less familiar than you on this code. Another question is whether we should consider resetting this.current to be START during teardownListeners()?